src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,48 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private Node<T> top;
+    private int size;
 
+    public void push(T value) {
+        top = new Node<>(top, value);
+        size++;
     }
 
     public T peek() {
-        return null;
+        if (isEmpty()) {
+            throw new EmptyStackException();
+        }
+        return top.item;
     }
 
     public T pop() {
-        return null;
+        if (isEmpty()) {
+            throw new EmptyStackException();
+        }
+        T value = peek();
+        top = top.prev;
+        size--;
+        return value;
     }
 
     public int size() {
-        return 0;
+        return size;
+    }
+
+    public boolean isEmpty() {
+        return size == 0;
+    }
+
+    private class Node<E> {
+        private E item;
+        private Node<E> prev;
+
+        Node(Node<E> prev, E element) {
+            this.item = element;
+            this.prev = prev;
+        }
     }
```

Comment:
It's recommended to document the private 'Node' class and its constructor to explain their roles within the stack structure.